Issue #5162 CDI embedded integration improvements

gregw opened the pull request:
Clean up CDI integration and documentation to better support embedded usage.
 + made listener public
 + added utility class for SCIs

gregw committed:
Issue #5162 CDI embedded integration improvements

Clean up CDI integration and documentation to better support embedded usage.
 + made listener public
 + added utility class for SCIs

gregw commented:
@janbartel @joakime can I get some feedback on this draft.

gregw commented on jetty-cdi/src/main/java/org/eclipse/jetty/cdi/CdiDecoratingListener.java:
```diff
@@ -24,13 +24,14 @@
  * A DecoratingListener that listens for "org.eclipse.jetty.cdi.decorator"
  */
-class CdiDecoratingListener extends DecoratingListener
+public class CdiDecoratingListener extends DecoratingListener
```
Made public so it can be directly used without SCI

gregw commented on jetty-cdi/src/main/java/org/eclipse/jetty/cdi/CdiDecoratingListener.java:
```diff
@@ -24,13 +24,14 @@
     public CdiDecoratingListener(ServletContextHandler contextHandler)
     {
         super(contextHandler, ATTRIBUTE);
+        contextHandler.setAttribute(CdiServletContainerInitializer.CDI_INTEGRATION_ATTRIBUTE, MODE);
```
Set the mode here, so direct usage still signals to weld the method used.

gregw commented on jetty-cdi/src/main/java/org/eclipse/jetty/cdi/CdiSpiDecorator.java:
```diff
@@ -65,7 +65,10 @@
+        context.setAttribute(CdiServletContainerInitializer.CDI_INTEGRATION_ATTRIBUTE, MODE);
         ClassLoader classLoader = _context.getClassLoader();
+        if (classLoader == null)
+            classLoader = this.getClass().getClassLoader();
```
make work when there is no classloader

gregw commented on jetty-cdi/src/main/java/org/eclipse/jetty/cdi/CdiSpiDecorator.java:
```diff
@@ -92,6 +95,35 @@ public CdiSpiDecorator(ServletContextHandler context) throws UnsupportedOperatio
+     */
+    protected boolean isKnownUndecoratable(String className)
+    {
+        return "org.jboss.weld.environment.servlet.Listener".equals(className);
```
This complex little dance is because it is too easy to install `CdiSpiDecorator` and then try to init Weld with it's listener, which will fail because `CdiSpiDecorator` will try to decorate it, but CDI is not enabled yet.     This soft dependency  breaks that chicken-egg problem.

gregw commented on jetty-plus/src/main/java/org/eclipse/jetty/plus/annotation/ContainerInitializer.java:
```diff
@@ -124,9 +124,6 @@ public void callStartup(WebAppContext context)
 
-            ClassLoader oldLoader = Thread.currentThread().getContextClassLoader();
-            Thread.currentThread().setContextClassLoader(context.getClassLoader());
-
```
This is no longer needed.  Always called with correct classloader set.

gregw commented on jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletContextHandler.java:
```diff
@@ -336,9 +338,15 @@ protected ServletHandler newServletHandler()
-        ServletContainerInitializerCaller sciBean = getBean(ServletContainerInitializerCaller.class);
-        if (sciBean != null)
-            sciBean.start();
+        for (ServletContainerInitializerCaller  sci : getBeans(ServletContainerInitializerCaller.class))
+        {
+            if (sci.isStopped())
+            {
+                sci.start();
+                if (isAuto(sci))
+                    manage(sci);
+            }
+        }
```
This allows for multiple SCI Callers.  It also correctly sets them as managed if this handler starts them.

gregw commented on jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletContextHandler.java:
```diff
@@ -1493,4 +1501,43 @@ public void destroy(Object o)
+     * {@link ServletContainerInitializerCaller} interface so that the SCI is correctly
+     * started if an instance of this class is added as a bean to a {@link ServletContextHandler}.
+     */
+    public static class Initializer extends AbstractLifeCycle implements ServletContainerInitializerCaller
+    {
+        private final ServletContextHandler _context;
+        private final ServletContainerInitializer _sci;
+        private final Set<Class<?>> _classes;
+
+        public Initializer(ServletContextHandler context, ServletContainerInitializer sci, Set<Class<?>> classes)
+        {
+            _context = context;
+            _sci = sci;
+            _classes = classes;
+        }
+
+        public Initializer(ServletContextHandler context, ServletContainerInitializer sci)
+        {
+            this(context, sci, Collections.emptySet());
+        }
+
+        @Override
+        protected void doStart() throws Exception
+        {
+            boolean oldExtended = _context.getServletContext().isExtendedListenerTypes();
+            try
+            {
+                _context.getServletContext().setExtendedListenerTypes(true);
+                _sci.onStartup(_classes, _context.getServletContext());
+            }
+            finally
+            {
+                _context.getServletContext().setExtendedListenerTypes(oldExtended);
+            }
+        }
+    }
```
This utility class avoids the need to create custom SCI callers.   The name is not perfect, but this name space is a bit messed up.  It does read well when used.

joakime replied:
If an SCI is being run on a ServletContextHandler, there's no expectations of bytecode scanning (IMO).

It would be sweet if we could just do ...

``` java
ServletContextHandler handler = new ServletContextHandler();
handler.addBean(new MyLibServletContextInitializer());
handler.addBean(new org.jboss.weld.environment.servlet.EnhancedListener()); // this is an SCI
server.setHandler(handler);
server.start(); // this loads the SCI's from the beans and uses them.
```

Just have ServletContextHandler be aware of SCI as beans and load / call them with no/empty classes.

Or, if this raw/naked SCI as bean concept isn't right, how about a specific call on ServletContextHandler for them?

``` java
ServletContextHandler handler = new ServletContextHandler();
handler.addInitializer(new MyLibServletContextInitializer(), Collections.emptySet());
handler.addInitializer(new org.jboss.weld.environment.servlet.EnhancedListener(), Collections.emptySet());
server.setHandler(handler);
server.start();
```

gregw replied:
@joakime if we were adding this mechanism from scratch, I agree that just adding the SCIs as beans would be simplest.
However, we already have the concept of a ServletContainerInitializerCaller set as a bean and it is used at least a little bit.... so adding a utility instance of that feels simpler than having two mechanisms.     I guess we could make `Initializer` private and check for SCI beans... so it would be a feeder to the existing mechanism.... torn.... @janbartel thoughts?

janbartel replied:
You can't do away with some kind of wrapper for the SCIs because the SCI.onStartup() method can have the list of classes passed in, so you need to remember somewhere what they are.

gregw replied:
Ah yes - naked SCIs only work if they have no classes to be passed in.  Whilst that will be rare for the usage intended, it still would mean another mechanism, so let's just keep it simple.

gregw commented on tests/test-distribution/src/test/java/org/eclipse/jetty/tests/distribution/CDITests.java:
```diff
@@ -61,7 +61,7 @@
             Arguments.of("weld", "cdi-spi", null), // Weld >= 3.1.2
             Arguments.of("weld", "decorate", null), // Weld >= 3.1.2
-            // TODO Arguments.of("weld", "cdi-decorate", null), // Weld >= 3.1.3
+            Arguments.of("weld", "cdi-decorate", null), // Weld >= 3.1.3
```
This test mode is now supported

gregw commented on tests/test-webapps/test-weld-cdi-webapp/src/test/java/org/eclipse/jetty/cdi/weld/EmbeddedWeldTest.java:
```diff
@@ -0,0 +1,241 @@
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsString;
+
+public class EmbeddedWeldTest
```
Added this unit test to test similar features as the webapp, but from embedded

gregw commented on tests/test-webapps/test-weld-cdi-webapp/src/test/java/org/eclipse/jetty/cdi/weld/EmbeddedWeldTest.java:
```diff
@@ -0,0 +1,241 @@
+        context.addFilter(ServerIDFilter.class, "/*", EnumSet.of(DispatcherType.REQUEST));
+
+        // Setup Jetty weld integration
+        switch (mode)
```
I don't think all these modes are recommended, but all variations are probably used somewhere so best to test to know if we break any usages.

gregw commented on tests/test-webapps/test-weld-cdi-webapp/src/test/java/org/eclipse/jetty/cdi/weld/EmbeddedWeldTest.java:
```diff
@@ -0,0 +1,241 @@
+        server.setHandler(webapp);
+
+        // Need the AnnotationConfiguration to detect SCIs
+        Configuration.ClassList.setServerDefault(server).addBefore(JettyWebXmlConfiguration.class.getName(),
+            AnnotationConfiguration.class.getName());
+
+        // Need to expose our SCI.  This is ugly could be made better in jetty-10 with a CdiConfiguration
+        webapp.getServerClasspathPattern().add("-" + CdiServletContainerInitializer.class.getName());
+        webapp.getSystemClasspathPattern().add(CdiServletContainerInitializer.class.getName());
+
```
We can add a CdiConfiguration to do this in 10.

gregw commented:
Hmm test harness approach doesn't work without IDE... refactoring...

gregw commented:
@WalkerWatch can you check the doco changes in this PR

gregw committed:
Issue #5162 CDI embedded integration improvements

Clean up CDI integration and documentation to better support embedded usage.
 + moved EmbeddedWeldTest to jetty-embedded

gregw committed:
Merge remote-tracking branch 'origin/jetty-9.4.x' into jetty-9.4.x-5162-Improve-Decorating-Listener-Examples

gregw committed:
fix javadoc

Signed-off-by: Greg Wilkins <gregw@webtide.com>

WalkerWatch reviewed (approved):
Documentation is fine with one caveat. We try to make use of [ventilated prose in asciidoc](https://asciidoctor.org/docs/asciidoc-recommended-practices/#one-sentence-per-line) as it makes maintenance and upkeep easier.

gregw committed:
Issue #5162 CDI embedded integration improvements

ventilated text

gregw committed:
Merge branch 'jetty-9.4.x-5162-Improve-Decorating-Listener-Examples' of github.com:eclipse/jetty.project into jetty-9.4.x-5162-Improve-Decorating-Listener-Examples

gregw committed:
Merge remote-tracking branch 'origin/jetty-9.4.x' into jetty-9.4.x-5162-Improve-Decorating-Listener-Examples

gregw committed:
fix test pom

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Fixed javadoc

gregw committed:
Merge branch 'jetty-9.4.x-5162-Improve-Decorating-Listener-Examples' of github.com:eclipse/jetty.project into jetty-9.4.x-5162-Improve-Decorating-Listener-Examples

gregw committed:
Fixed javadoc

gregw committed:
Merge remote-tracking branch 'origin/jetty-9.4.x' into jetty-9.4.x-5162-Improve-Decorating-Listener-Examples

gregw committed:
Issue #5162 CDI embedded integration improvements

Moved tests to jetty-cdi to avoid consequences to other tests in embedded

janbartel commented on jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletContextHandler.java:
```diff
@@ -749,13 +757,13 @@ public static ServletContextHandler getServletContextHandler(ServletContext cont
     public static class JspPropertyGroup implements JspPropertyGroupDescriptor
     {
-        private List<String> _urlPatterns = new ArrayList<>();
+        private final List<String> _urlPatterns = new ArrayList<>();
```
Again with the extraneous changes!

gregw replied:
yeah - you have to be a little bad sometimes :)

janbartel commented on jetty-util/src/main/java/org/eclipse/jetty/util/component/ContainerLifeCycle.java:
```diff
@@ -915,11 +915,6 @@ public void updateBeans(Object[] oldBeans, final Object[] newBeans)
         }
     }
 
-    /**
```
Why is the comment removed?

gregw replied:
You don't need to comment overridden methods unless you need to add impl notes.

janbartel commented on jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletContextHandler.java:
```diff
@@ -1493,4 +1501,43 @@ public void destroy(Object o)
+     * {@link ServletContainerInitializerCaller} interface so that the SCI is correctly
+     * started if an instance of this class is added as a bean to a {@link ServletContextHandler}.
+     */
+    public static class Initializer extends AbstractLifeCycle implements ServletContainerInitializerCaller
```
I don't see any refactoring of the existing ContainerInitializer and ServletContainerInitializersStarter classes used by webapps - I would think that we could extend this ServletContextHandler.Initializer class, or have the other classes extend it. I don't like having 2 completely separate mechanisms.

gregw replied:
We have a single mechanism based on a single interface, `ServletContainerInitializerCaller`, that has two key implementations:
 1. The webapp impl that has a single caller that invokes all the SCIs it has discovered in the order that it wants.
 2. The Initializer impl that is just a convenience class to create your own custom caller.

To make the webapp impl use the Initializer, then it would have to make all discovered SCIs individually wrapped as Initializers and get into the business of sorting beans on the context....  That is doable, but I'm not sure it is worth it for the bugs that it will introduce.  I really don't want to mess with webapps SCI handling if it is not broken.

janbartel replied:
But I thought that the ordering on the beans is that they are called in the order they are added? If so, then we can just have AnnotationConfiguration add them in the order that is established by applying any ordering: currently it just adds the ContainerInitializers to a list in the order determined by the ordering, so I don't see the difference between adding it to that list, and adding it to the  list of beans.

And I think the refactoring I'm talking about should be done in 10, not 9.

gregw replied:
Yes it is totally doable... but I just don't want to mess with something that isn't broken.... maybe 10.1?

gregw committed:
trailing new line

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw commented:
@janbartel What are the show stoppers for you in this PR?

gregw commented:
@janbartel @joakime nudge!

janbartel commented on jetty-cdi/src/main/java/org/eclipse/jetty/cdi/CdiSpiDecorator.java:
```diff
@@ -92,6 +95,35 @@ public CdiSpiDecorator(ServletContextHandler context) throws UnsupportedOperatio
+    {
+        if (Object.class == clazz)
+            return true;
+        if (isKnownUndecoratable(clazz.getName()))
```
Not at all fond of the name isKnownDecoratable(). How about instead you remove isKnownDecoratable() and add a field list of _undecoratables, a setter for them, and in isDecoratable() you just check the list? This might also be a good idea in case CDI adds some other classes that can't be decorated?

janbartel commented on jetty-documentation/src/main/asciidoc/development/frameworks/cdi.adoc:
```diff
@@ -0,0 +1,119 @@
+
+However, depending on the exact configuration of the embedded server, either or both steps may not be required as Servlet Container Initializers may be discovered.
+
+The details for embedding CDI will be explain in the Embedded Jetty with Weld section, but can be adapted to other CDI frameworks.
```
Need a link here?

janbartel reviewed (approved):
So I don't have any showstoppers, but I would like you to take a look at my suggestion for the isKnownDecoratable() method.

gregw committed:
updates from review

Signed-off-by: Greg Wilkins <gregw@webtide.com>

janbartel commented on jetty-cdi/src/main/java/org/eclipse/jetty/cdi/CdiSpiDecorator.java:
```diff
@@ -92,6 +100,54 @@ public CdiSpiDecorator(ServletContextHandler context) throws UnsupportedOperatio
+     * @param undecorated The set of class names that will not be decorated.
+     * @see #isDecoratable(Class)
+     */
+    public void setUndecoratedClasses(Set<String> undecorated)
```
I meant "setUndecoratableClasses". Undecorated makes it sound like they haven't been decorated yet.

gregw replied:
@janb I saw that, but I think Undecorated also works, is less of a mouthful and eitherway you need to check with the javadoc to really know what it means.

gregw committed:
Feedback from review

janbartel reviewed (approved):
LGTM
